## Gather

In [831]:
import pandas as pd
import numpy as py
import requests as r
import tweepy as tw
import json as js

In [832]:
df_1 = pd.read_csv("twitter-archive-enhanced.csv")

In [833]:
# Twitter Archive - initial exploration but here still in gathering phase. 
df_1.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [834]:
df_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [835]:
#Image Predictions - gathering second portion of data.
r.get('https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv')

<Response [200]>

In [836]:
# Reading the tab delimited file. 
df_2 = pd.read_csv('image-predictions.tsv',  delimiter='\t')

In [837]:
df_2.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [838]:
df_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [839]:
# Section commented out to show code and keep keys hidden
# Twitter API - gathering data from Twitter.
# import tweepy
# from tweepy import OAuthHandler
# import json
# from timeit import default_timer as timer

# Query Twitter API for each tweet in the Twitter archive and save JSON in a text file
# These are hidden to comply with Twitter's API terms and conditions
# consumer_key = 'HIDDEN'
# consumer_secret = 'HIDDEN'
# access_token = 'HIDDEN'
# access_secret = 'HIDDEN'

# auth = OAuthHandler(consumer_key, consumer_secret)
# auth.set_access_token(access_token, access_secret)

# api = tweepy.API(auth, wait_on_rate_limit=True)

# tweet_ids = df_1.tweet_id.values
# len(tweet_ids)

# Query Twitter's API for JSON data for each tweet ID in the Twitter archive
# count = 0
# fails_dict = {}
# start = timer()
# Save each tweet's returned JSON as a new line in a .txt file
# with open('tweet_json_2.txt', 'w') as outfile:
    # This loop will likely take 20-30 minutes to run because of Twitter's rate limit
#    for tweet_id in tweet_ids:
#       count += 1
        # string check
#        print(type(tweet_id))
#        print(str(count) + ": " + str(tweet_id))
#        try:
#            tweet = api.get_status(tweet_id, tweet_mode='extended')
#            print("Success")
#            json.dump(tweet._json, outfile)
#            outfile.write('\n')
#        except tweepy.TweepError as e:
#            print("Fail")
#            fails_dict[tweet_id] = e
#            pass
        
# end = timer()
# print(end - start)
# print(fails_dict)


In [840]:
# Loading data into dataframe, commenting out with API keys hidden
# tweets = []
# with open ('tweet_json_2.txt', 'r') as file:
#        for line in file:
#            tweet = json.loads(line)
#            tweet_id = tweet['id']
#            retweet_count = tweet['retweet_count']
#            fav_count = tweet['favorite_count']
#            user_count = tweet['user']['followers_count']
#            tweets.append({'tweet_id':tweet_id,
#                          'retweet_count': retweet_count,
#                          'favorite_count': fav_count,
#                          'user_count':user_count})
# df_3 = pd.DataFrame(data=tweets)


In [841]:
# Save to csv, commenting out with API keys hidden 
# df_3.to_csv('tweets_from_api.csv', index = False) 

In [842]:
# Open from csv, loads data from functioning API pull 
df_3b = pd.read_csv('tweets_from_api.csv')

In [843]:
df_3b.head()

tweet_id  retweet_count  favorite_count  user_count
0  892420643555336193           7142           34350     9099508
1  892177421306343426           5374           29779     9099508
2  891815181378084864           3541           22402     9099508
3  891689557279858688           7336           37503     9099508
4  891327558926688256           7888           35830     9099508

In [844]:
df_3b.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2328 entries, 0 to 2327
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   tweet_id        2328 non-null   int64
 1   retweet_count   2328 non-null   int64
 2   favorite_count  2328 non-null   int64
 3   user_count      2328 non-null   int64
dtypes: int64(4)
memory usage: 72.9 KB


## Assess

In [845]:
# Twitter Archive 
# Visual exploration first; see additional notes in the wrangle_report. 
df_1

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
...                  ...                    ...                  ...   
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
0     2017-08-01 16:23:56 +0000   
1     2017-08-01 00:17:27 +0000   
2     2017-07-31 00:18:03 +0000   
3     2017-07-30 15:58:51 +0000   
4     2017-07-29 16:00:24 +0000   
...                         ...   
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
0     This is Phineas. He's a mystical boy. Only eve...                  NaN   
1     This is Tilly. She's just checking pup on you....                  NaN   
2     This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3     This is Darla. She commenced a snooze mid meal...                  NaN   
4     This is Franklin. He would like you to stop ca...                  NaN   
...                                                 ...                  ...   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
0                          NaN                        NaN   
1                          NaN                        NaN   
2                          NaN                        NaN   
3                          NaN                        NaN   
4                          NaN                        NaN   
...                        ...                        ...   
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
0     https://twitter.com/dog_rates/status/892420643...                13   
1     https://twitter.com/dog_rates/status/892177421...                13   
2     https://twitter.com/dog_rates/status/891815181...                12   
3     https://twitt

In [846]:
# Image Predictions
df_2

tweet_id                                          jpg_url  \
0     666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
...                  ...                                              ...   
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num                      p1   p1_conf  p1_dog                  p2  \
0           1  Welsh_springer_spaniel  0.465074    True              collie   
1           1                 redbone  0.506826    True  miniature_pinscher   
2           1         German_shepherd  0.596461    True            malinois   
3           1     Rhodesian_ridgeback  0.408143    True             redbone   
4           1      miniature_pinscher  0.560311    True          Rottweiler   
...       ...                     ...       ...     ...                 ...   
2070        2                  basset  0.555712    True    English_springer   
2071        1             paper_towel  0.170278   False  Labrador_retriever   
2072        1               Chihuahua  0.716012    True            malamute   
2073        1               Chihuahua  0.323581    True            Pekinese   
2074        1                  orange  0.097049   False               bagel   

       p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0     0.156665    True            Shetland_sheepdog  0.061428    True  
1     0.074192    True          Rhodesian_ridgeback  0.072010    True  
2     0.138584    True                   bloodhound  0.116197    True  
3     0.360687    True           miniature_pinscher  0.222752    True  
4     0.243682    True                     Doberman  0.154629    True  
...        ...     ...                          ...       ...     ...  
2070  0.225770    True  German_short-haired_pointer  0.175219    True  
2071  0.168086    True                      spatula  0.040836   False  
2072  0.078253    True                       kelpie  0.031379    True  
2073  0.090647    True                     papillon  0.068957    True  
2074  0.085851   False                       banana  0.076110   False  

[2075 rows x 12 columns]

In [847]:
# Twitter API
df_3b

tweet_id  retweet_count  favorite_count  user_count
0     892420643555336193           7142           34350     9099508
1     892177421306343426           5374           29779     9099508
2     891815181378084864           3541           22402     9099508
3     891689557279858688           7336           37503     9099508
4     891327558926688256           7888           35830     9099508
...                  ...            ...             ...         ...
2323  666049248165822465             37              91     9099499
2324  666044226329800704            117             251     9099499
2325  666033412701032449             36             102     9099499
2326  666029285002620928             39             115     9099499
2327  666020888022790149            432            2316     9099499

[2328 rows x 4 columns]

In [848]:
# Programatic assessment; see list at end for cleaning observations. 
df_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [849]:
df_1['rating_numerator'].value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
420       2
0         2
15        2
75        2
80        1
20        1
24        1
26        1
44        1
50        1
60        1
165       1
84        1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
17        1
27        1
45        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

In [850]:
df_1['rating_denominator'].value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

In [851]:
df_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [852]:
df_3b.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2328 entries, 0 to 2327
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   tweet_id        2328 non-null   int64
 1   retweet_count   2328 non-null   int64
 2   favorite_count  2328 non-null   int64
 3   user_count      2328 non-null   int64
dtypes: int64(4)
memory usage: 72.9 KB


In [853]:
df_3b.describe().apply(lambda s: s.apply('{0:.1f}'.format))

tweet_id retweet_count favorite_count user_count
count                2328.0        2328.0         2328.0     2328.0
mean   741840270868517248.0        2516.5         7164.2  9099503.0
std     68231434753658864.0        4260.9        11129.2       44.4
min    666020888022790144.0           1.0            0.0  9099110.0
25%    678197428948985856.0         507.0         1244.2  9099508.0
50%    718038209626820608.0        1170.5         3096.5  9099509.0
75%    798667311000495104.0        2913.2         8741.8  9099511.0
max    892420643555336192.0       72132.0       147108.0  9099573.0

## Assessment Summary

#### Quality

#### Twitter Archive 'df_1' 
- tweet_id is an int64 field and needs to be a string
- in_reply_to_status_id and in_reply_to_user_id have 78 non-null values that can be removed
- retweeted_status_id contains 181 non-null items that can be removed
- timestamp could be more helpful if split into date and time
- many numerator ratings are incorrect; defined as greater than 15 in the numerator
- many denominator values aren't 10; since 10 is always the scale

#### Image Predictions 'df_2'
- tweet_id is an int64 field and needs to be a string
- capitalizations are not consistent
- confidence format of 6-digit decimal is hard to read, round to 2 digits

#### Twitter API 'df_3b'
- tweet_id is an int64 field and needs to be a string

#### Tidiness

#### Twitter Archive 'df_1' 
- some columns will be left with null data: 
    in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, retweeted_status_timestamp
- the doggos, floofer, pupper and puppo fields can be combined into one column

#### Twitter API 'df_3'
- user_count can be dropped

#### Combined
- data can be joined on tweet_id to combine 3 tables into 1

## Clean

In [854]:
# Creating copies of each file.
df_1.to_csv('df_1_copy.csv', index = False) 
df_2.to_csv('df_2_copy.csv', index = False) 
df_3b.to_csv('df_3_copy.csv', index = False)

In [855]:
# Starting with fresh dataframes from these saved csv files
df_1c = pd.read_csv('df_1_copy.csv')
df_2c = pd.read_csv('df_2_copy.csv')
df_3c = pd.read_csv('df_3_copy.csv')

### Quality
##### Define
In Twitter Archive 'df_1c', convert tweet_id to string 

##### Code

In [856]:
df_1c['tweet_id'] = df_1c['tweet_id'].astype(str)

##### Test

In [858]:
df_1c.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   object 
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

##### Define
Remove tweets with non-null values in_reply_to_status_id

##### Code

In [859]:
df_1c = df_1c[df_1c.in_reply_to_status_id.isnull()]

##### Test

In [860]:
df_1c.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2278 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2278 non-null   object 
 1   in_reply_to_status_id       0 non-null      float64
 2   in_reply_to_user_id         0 non-null      float64
 3   timestamp                   2278 non-null   object 
 4   source                      2278 non-null   object 
 5   text                        2278 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2274 non-null   object 
 10  rating_numerator            2278 non-null   int64  
 11  rating_denominator          2278 non-null   int64  
 12  name                        2278 non-null   object 
 13  doggo                       2278 

##### Define
Remove retweets with non-null values retweeted_status_id

##### Code

In [861]:
df_1c = df_1c[df_1c.retweeted_status_id.isnull()]

##### Test

In [862]:
df_1c.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2097 non-null   object 
 1   in_reply_to_status_id       0 non-null      float64
 2   in_reply_to_user_id         0 non-null      float64
 3   timestamp                   2097 non-null   object 
 4   source                      2097 non-null   object 
 5   text                        2097 non-null   object 
 6   retweeted_status_id         0 non-null      float64
 7   retweeted_status_user_id    0 non-null      float64
 8   retweeted_status_timestamp  0 non-null      object 
 9   expanded_urls               2094 non-null   object 
 10  rating_numerator            2097 non-null   int64  
 11  rating_denominator          2097 non-null   int64  
 12  name                        2097 non-null   object 
 13  doggo                       2097 

##### Define
Split out date and time from the timestamp field.

##### Code

In [863]:
df_1c['date'] = pd.to_datetime(df_1c['timestamp']).dt.date
df_1c['time'] = pd.to_datetime(df_1c['timestamp']).dt.time


##### Test

In [864]:
df_1c.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 19 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2097 non-null   object 
 1   in_reply_to_status_id       0 non-null      float64
 2   in_reply_to_user_id         0 non-null      float64
 3   timestamp                   2097 non-null   object 
 4   source                      2097 non-null   object 
 5   text                        2097 non-null   object 
 6   retweeted_status_id         0 non-null      float64
 7   retweeted_status_user_id    0 non-null      float64
 8   retweeted_status_timestamp  0 non-null      object 
 9   expanded_urls               2094 non-null   object 
 10  rating_numerator            2097 non-null   int64  
 11  rating_denominator          2097 non-null   int64  
 12  name                        2097 non-null   object 
 13  doggo                       2097 

In [865]:
df_1c.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo        date  \
0                  10   Phineas  None    None   None  None  2017-08-01   
1                  10     Tilly  None    None   None  None  2017-08-01   
2                  10    Archie  None    None   None  None  2017-07-31   
3                  10     Darla  None    None   None  None  2017-07-30   
4                  10  Franklin  None    None   None  None  2017-07-29   

       time  
0  16:23:56  
1  00:17:27  
2  00:18:03  
3  15:58:51  
4  16:00:24

##### Define
Remove tweets with incorrect ratings in the numerator; with errors defined as 15 or greater

Also, rating of 0 seems to be an error as well

##### Code

In [867]:
df_1c = df_1c[df_1c['rating_numerator'].between(1,14)]

##### Test

In [868]:
df_1c['rating_numerator'].value_counts()

12    486
10    436
11    413
13    287
9     153
8      98
7      52
14     38
5      34
6      32
3      19
4      16
2       9
1       5
Name: rating_numerator, dtype: int64

In [869]:
df_1c['rating_denominator'].value_counts()

10    2074
11       2
20       1
2        1
Name: rating_denominator, dtype: int64

##### Define
Correct all demoninators to 10

##### Code

In [870]:
df_1c['rating_denominator'] = 10

##### Test

In [871]:
df_1c['rating_denominator'].value_counts()

10    2078
Name: rating_denominator, dtype: int64

##### Define
In Image Predictions 'df_2c', convert tweet_id to string 

##### Code

In [872]:
df_2c['tweet_id'] = df_2c['tweet_id'].astype(str)

##### Test

In [873]:
df_2c.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   object 
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(1), object(5)
memory usage: 152.1+ KB


##### Define
Remove the underscores in the dog names

Then make them all title case 

##### Code

In [874]:
df_2c['p1']=df_2c['p1'].str.replace('_', ' ')
df_2c['p2']=df_2c['p2'].str.replace('_', ' ')
df_2c['p3']=df_2c['p3'].str.replace('_', ' ')

##### Test

In [875]:
df_2c.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh springer spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature pinscher   
2        1         German shepherd  0.596461    True            malinois   
3        1     Rhodesian ridgeback  0.408143    True             redbone   
4        1      miniature pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland sheepdog  0.061428    True  
1  0.074192    True  Rhodesian ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

##### Code

In [876]:
df_2c['p1']=df_2c['p1'].str.title()
df_2c['p2']=df_2c['p2'].str.title()
df_2c['p3']=df_2c['p3'].str.title()

##### Test

In [877]:
df_2c.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh Springer Spaniel  0.465074    True              Collie   
1        1                 Redbone  0.506826    True  Miniature Pinscher   
2        1         German Shepherd  0.596461    True            Malinois   
3        1     Rhodesian Ridgeback  0.408143    True             Redbone   
4        1      Miniature Pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland Sheepdog  0.061428    True  
1  0.074192    True  Rhodesian Ridgeback  0.072010    True  
2  0.138584    True           Bloodhound  0.116197    True  
3  0.360687    True   Miniature Pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [878]:
df_2c['p1_conf']=df_2c['p1_conf'].round(2)
df_2c['p2_conf']=df_2c['p2_conf'].round(2)
df_2c['p3_conf']=df_2c['p3_conf'].round(2)

##### Test

In [879]:
df_2c.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1  p1_conf  p1_dog                  p2  \
0        1  Welsh Springer Spaniel     0.47    True              Collie   
1        1                 Redbone     0.51    True  Miniature Pinscher   
2        1         German Shepherd     0.60    True            Malinois   
3        1     Rhodesian Ridgeback     0.41    True             Redbone   
4        1      Miniature Pinscher     0.56    True          Rottweiler   

   p2_conf  p2_dog                   p3  p3_conf  p3_dog  
0     0.16    True    Shetland Sheepdog     0.06    True  
1     0.07    True  Rhodesian Ridgeback     0.07    True  
2     0.14    True           Bloodhound     0.12    True  
3     0.36    True   Miniature Pinscher     0.22    True  
4     0.24    True             Doberman     0.15    True

##### Define
In Twitter API 'df_3c', convert tweet_id to string 

##### Code

In [880]:
df_3c['tweet_id'] = df_3c['tweet_id'].astype(str)

##### Test

In [881]:
df_3c.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2328 entries, 0 to 2327
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   tweet_id        2328 non-null   object
 1   retweet_count   2328 non-null   int64 
 2   favorite_count  2328 non-null   int64 
 3   user_count      2328 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 72.9+ KB


#### Tidiness
##### Define
In Twitter Archive 'df_1c', drop columns with only null values, and the timestamp column. 

##### Code

In [882]:
df_1c.drop(df_1c.columns[[1, 2, 3, 6, 7, 8]], axis = 1, inplace = True)

##### Test

In [883]:
df_1c.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2078 entries, 0 to 2355
Data columns (total 13 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   tweet_id            2078 non-null   object
 1   source              2078 non-null   object
 2   text                2078 non-null   object
 3   expanded_urls       2075 non-null   object
 4   rating_numerator    2078 non-null   int64 
 5   rating_denominator  2078 non-null   int64 
 6   name                2078 non-null   object
 7   doggo               2078 non-null   object
 8   floofer             2078 non-null   object
 9   pupper              2078 non-null   object
 10  puppo               2078 non-null   object
 11  date                2078 non-null   object
 12  time                2078 non-null   object
dtypes: int64(2), object(11)
memory usage: 227.3+ KB


##### Define
In Twitter Archive 'df_1c', combine the doggo, floofer, pupper and puppo fields into 1 column

##### Code

In [884]:
df_1c = pd.melt(df_1c, id_vars=['tweet_id', 
                                'source',
                                'text',
                                'expanded_urls',
                                'rating_numerator',
                                'rating_denominator',
                                'name',
                               'date',
                               'time'],
                var_name = 'dogs', value_name='dogs_stage')
df_1c = df_1c.drop('dogs',1)

##### Test

In [885]:
df_1c.head()

tweet_id                                             source  \
0  892420643555336193  <a href="http://twitter.com/download/iphone" r...   
1  892177421306343426  <a href="http://twitter.com/download/iphone" r...   
2  891815181378084864  <a href="http://twitter.com/download/iphone" r...   
3  891689557279858688  <a href="http://twitter.com/download/iphone" r...   
4  891327558926688256  <a href="http://twitter.com/download/iphone" r...   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   
1  This is Tilly. She's just checking pup on you....   
2  This is Archie. He is a rare Norwegian Pouncin...   
3  This is Darla. She commenced a snooze mid meal...   
4  This is Franklin. He would like you to stop ca...   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name        date      time dogs_stage  
0                  10   Phineas  2017-08-01  16:23:56       None  
1                  10     Tilly  2017-08-01  00:17:27       None  
2                  10    Archie  2017-07-31  00:18:03       None  
3                  10     Darla  2017-07-30  15:58:51       None  
4                  10  Franklin  2017-07-29  16:00:24       None

In [886]:
df_1c['dogs_stage'].value_counts()

None       7966
pupper      229
doggo        83
puppo        24
floofer      10
Name: dogs_stage, dtype: int64

#### Tidiness
##### Define
In Twitter API 'df_3c,, drop the user_count column  

In [887]:
df_3c.drop(df_3c.columns[3], axis = 1, inplace = True)

##### Test

In [888]:
df_3c.head()

tweet_id  retweet_count  favorite_count
0  892420643555336193           7142           34350
1  892177421306343426           5374           29779
2  891815181378084864           3541           22402
3  891689557279858688           7336           37503
4  891327558926688256           7888           35830

##### Define
Merge the 3 tables into 1 dataset on tweet_id

In [889]:
twitter_archive_master_1 = pd.merge(df_1c, df_2c,
                            on=['tweet_id'], how='inner')

In [890]:
twitter_archive_master_1.head()

tweet_id                                             source  \
0  892420643555336193  <a href="http://twitter.com/download/iphone" r...   
1  892420643555336193  <a href="http://twitter.com/download/iphone" r...   
2  892420643555336193  <a href="http://twitter.com/download/iphone" r...   
3  892420643555336193  <a href="http://twitter.com/download/iphone" r...   
4  892177421306343426  <a href="http://twitter.com/download/iphone" r...   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   
1  This is Phineas. He's a mystical boy. Only eve...   
2  This is Phineas. He's a mystical boy. Only eve...   
3  This is Phineas. He's a mystical boy. Only eve...   
4  This is Tilly. She's just checking pup on you....   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892420643...                13   
2  https://twitter.com/dog_rates/status/892420643...                13   
3  https://twitter.com/dog_rates/status/892420643...                13   
4  https://twitter.com/dog_rates/status/892177421...                13   

   rating_denominator     name        date      time dogs_stage  ... img_num  \
0                  10  Phineas  2017-08-01  16:23:56       None  ...       1   
1                  10  Phineas  2017-08-01  16:23:56       None  ...       1   
2                  10  Phineas  2017-08-01  16:23:56       None  ...       1   
3                  10  Phineas  2017-08-01  16:23:56       None  ...       1   
4                  10    Tilly  2017-08-01  00:17:27       None  ...       1   

          p1 p1_conf  p1_dog        p2 p2_conf  p2_dog        p3 p3_conf  \
0     Orange    0.10   False     Bagel    0.09   False    Banana    0.08   
1     Orange    0.10   False     Bagel    0.09   False    Banana    0.08   
2     Orange    0.10   False     Bagel    0.09   False    Banana    0.08   
3     Orange    0.10   False     Bagel    0.09   False    Banana    0.08   
4  Chihuahua    0.32    True  Pekinese    0.09    True  Papillon    0.07   

   p3_dog  
0   False  
1   False  
2   False  
3   False  
4    True  

[5 rows x 21 columns]

In [891]:
twitter_archive_master_final = pd.merge(twitter_archive_master_1, df_3c,
                            on=['tweet_id'], how='inner')

In [820]:
twitter_archive_master_final.head()

tweet_id                                             source  \
0  892420643555336193  <a href="http://twitter.com/download/iphone" r...   
1  892420643555336193  <a href="http://twitter.com/download/iphone" r...   
2  892420643555336193  <a href="http://twitter.com/download/iphone" r...   
3  892420643555336193  <a href="http://twitter.com/download/iphone" r...   
4  892177421306343426  <a href="http://twitter.com/download/iphone" r...   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   
1  This is Phineas. He's a mystical boy. Only eve...   
2  This is Phineas. He's a mystical boy. Only eve...   
3  This is Phineas. He's a mystical boy. Only eve...   
4  This is Tilly. She's just checking pup on you....   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892420643...                13   
2  https://twitter.com/dog_rates/status/892420643...                13   
3  https://twitter.com/dog_rates/status/892420643...                13   
4  https://twitter.com/dog_rates/status/892177421...                13   

   rating_denominator     name        date      time dogs_stage  ... p1_conf  \
0                  10  Phineas  2017-08-01  16:23:56       None  ...    0.10   
1                  10  Phineas  2017-08-01  16:23:56       None  ...    0.10   
2                  10  Phineas  2017-08-01  16:23:56       None  ...    0.10   
3                  10  Phineas  2017-08-01  16:23:56       None  ...    0.10   
4                  10    Tilly  2017-08-01  00:17:27       None  ...    0.32   

   p1_dog        p2  p2_conf  p2_dog        p3  p3_conf  p3_dog retweet_count  \
0   False     Bagel     0.09   False    Banana     0.08   False          7142   
1   False     Bagel     0.09   False    Banana     0.08   False          7142   
2   False     Bagel     0.09   False    Banana     0.08   False          7142   
3   False     Bagel     0.09   False    Banana     0.08   False          7142   
4    True  Pekinese     0.09    True  Papillon     0.07    True          5374   

   favorite_count  
0           34350  
1           34350  
2           34350  
3           34350  
4           29779  

[5 rows x 23 columns]

In [892]:
twitter_archive_master_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7776 entries, 0 to 7775
Data columns (total 23 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   tweet_id            7776 non-null   object 
 1   source              7776 non-null   object 
 2   text                7776 non-null   object 
 3   expanded_urls       7776 non-null   object 
 4   rating_numerator    7776 non-null   int64  
 5   rating_denominator  7776 non-null   int64  
 6   name                7776 non-null   object 
 7   date                7776 non-null   object 
 8   time                7776 non-null   object 
 9   dogs_stage          7776 non-null   object 
 10  jpg_url             7776 non-null   object 
 11  img_num             7776 non-null   int64  
 12  p1                  7776 non-null   object 
 13  p1_conf             7776 non-null   float64
 14  p1_dog              7776 non-null   bool   
 15  p2                  7776 non-null   object 
 16  p2_con

In [893]:
sum(twitter_archive_master_final.duplicated())

5521

In [894]:
twitter_archive_master_final = twitter_archive_master_final.drop_duplicates()

In [828]:
twitter_archive_master_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2255 entries, 0 to 7772
Data columns (total 23 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   tweet_id            2255 non-null   object 
 1   source              2255 non-null   object 
 2   text                2255 non-null   object 
 3   expanded_urls       2255 non-null   object 
 4   rating_numerator    2255 non-null   int64  
 5   rating_denominator  2255 non-null   int64  
 6   name                2255 non-null   object 
 7   date                2255 non-null   object 
 8   time                2255 non-null   object 
 9   dogs_stage          2255 non-null   object 
 10  jpg_url             2255 non-null   object 
 11  img_num             2255 non-null   int64  
 12  p1                  2255 non-null   object 
 13  p1_conf             2255 non-null   float64
 14  p1_dog              2255 non-null   bool   
 15  p2                  2255 non-null   object 
 16  p2_con

In [895]:
# Saving to csv for analysis in Tableau
twitter_archive_master_final.to_csv('twitter_archive_master_final.csv', index = False) 